plugins/cli/__init__.py

```diff
@@ -22,234 +23,548 @@
+        """
         lib.connection.Stream.__init__(self, sock, source)
+        self.logger = logging.getLogger(__name__)
         self.source = source
         self.updates_allowed = updates
         self.sh = smarthome
-        self.push("SmartHome.py v{0}\n".format(self.sh.version))
-        self.push("Enter 'help' for a list of available commands.\n")
-        self.push("> ")
+        self.hashed_password = hashed_password
+        self.commands = commands
+        self.push("SmartHomeNG v{0}\n".format(self.sh.version))
+        self.push("Password: ")
+
+        # Inform the client that we will echo what has been entered. He won't do this from now on.
+        # As we don't do this, too, the entered hashed_password will not be shown ...
+        self.send(bytearray([0xFF, 0xFB, 0x01]))  # IAC WILL ECHO
+        self.__wait_for_password = True
 
     def push(self, data):
+        """
+        Send data to client
+        :param data: String to send
+        """
         self.send(data.encode())
 
     def found_terminator(self, data):
+        """
+        Received data and found terminator (newline) in data
+        :param data: Received data up to terminator
+        """
+
+        # Every three bytes starting with 0xFF are a command. At the moment they will only occur at the start of the response, after we
+        # sent the IAC WILL ECHO/IAC WONT ECHO commands. They contain the confirmation of the commands (IAC DO ECHO: 0xFF 0xFD 0x01 or
+        # IAC DONT ECHO: 0xFF 0xFE 0x01) We will simple ignore these at the moment.
+        if data[0] == 0xFF:
+            data = data[3:]
+
         cmd = data.decode().strip()
-        if cmd.startswith('ls'):
-            self.push("Items:\n======\n")
-            self.ls(cmd.lstrip('ls').strip(), '*' in cmd or ':' in cmd)
-        elif cmd == 'la':
-            self.la()
-        elif cmd == 'lo':
-            self.lo()
-        elif cmd == 'll':
-            self.lo()
-        elif cmd == 'lt':
-            self.lt()
-        elif cmd == 'cl':
-            self.cl()
-        elif cmd.startswith('update ') or cmd.startswith('up '):
-            self.update(cmd.lstrip('update').strip())
-        elif cmd.startswith('dump'):
-            self.dump(cmd.lstrip('dump').strip(), '*' in cmd or ':' in cmd)
-        elif cmd.startswith('tr'):
-            self.tr(cmd.lstrip('tr').strip())
-        elif cmd.startswith('rl'):
-            self.rl(cmd.lstrip('rl').strip())
-        elif cmd.startswith('rr'):
-            self.rr(cmd.lstrip('rr').strip())
-        elif cmd == 'rt' or cmd == 'runtime':
-            self.rt()
-        elif cmd == 'help' or cmd == 'h':
-            self.usage()
-        elif cmd in ('quit', 'q', 'exit', 'x'):
+
+        if self.__wait_for_password:
+            # We are waiting for a password. So we take the entered command as password
+            if Utils.create_hash(cmd) == self.hashed_password.lower():
```

Comment:
You are right. It should be a method in utils.py. Not because the ".lower()" part might be forgotten by someone else, but because other checks could be made in there as well. E.g.:

- Empty password -> return "not matching"
- Password shorter than [insert min length here] characters -> return "not matching"
- ...